File: matching-engine/src/main/java/com/titanx/simulation/OrderGenerator.java
package com.titanx.simulation;

import com.titanx.engine.MatchingEngine;
import com.titanx.model.*;
import com.titanx.risk.CircuitBreaker;
import com.titanx.risk.RiskEngine;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;
import java.util.random.RandomGenerator;

/**
 * ============================================================
 * TitanX OrderGenerator — Load Testing Simulation
 * ============================================================
 *
 * Simulates realistic order flow to measure throughput and latency.
 *
 * Scenario: 10 market makers + 90 retail traders, all sending orders
 * to a single AAPL order book.
 *
 * Output:
 *   - Throughput: orders/second
 *   - Latency: p50, p95, p99 in microseconds
 *   - Fill rate: % of orders that resulted in at least one trade
 *   - Risk rejection rate
 *
 * Run: java -cp titanx-engine.jar com.titanx.simulation.OrderGenerator
 *
 * Target: 100,000 orders/minute (1,667/second) on commodity hardware.
 */
public class OrderGenerator {

    private static final String SYMBOL = "AAPL";
    private static final int SIMULATION_SECONDS = 30;
    private static final int THREADS = 20;
    private static final int ORDERS_PER_THREAD_PER_SECOND = 50;

    // Simulated market makers seed the book with resting orders
    private static final int MARKET_MAKER_ACCOUNTS = 10;

    public static void main(String[] args) throws InterruptedException {
        System.out.println("=================================================");
        System.out.println("  TitanX Load Simulation — Goldman Sachs Interview");
        System.out.println("=================================================");
        System.out.printf("  Threads: %d | Duration: %ds | Target: ~%d orders/sec%n%n",
                THREADS, SIMULATION_SECONDS, THREADS * ORDERS_PER_THREAD_PER_SECOND);

        CircuitBreaker cb = new CircuitBreaker();
        RiskEngine riskEngine = new RiskEngine(cb);
        MatchingEngine engine = new MatchingEngine(riskEngine);

        // Seed the order book with market maker quotes
        seedMarketMakerOrders(engine);

        // Metrics
        AtomicLong totalOrders  = new AtomicLong();
        AtomicLong totalFills   = new AtomicLong();
        AtomicLong totalLatency = new AtomicLong();
        List<Long> latencySamples = Collections.synchronizedList(new ArrayList<>());

        ExecutorService pool = Executors.newFixedThreadPool(THREADS);
        long startTime = System.currentTimeMillis();

        for (int t = 0; t < THREADS; t++) {
            final int threadId = t;
            pool.submit(() -> runTraderThread(threadId, engine, SIMULATION_SECONDS,
                    ORDERS_PER_THREAD_PER_SECOND, totalOrders, totalFills, latencySamples));
        }

        pool.shutdown();
        pool.awaitTermination(SIMULATION_SECONDS + 5, TimeUnit.SECONDS);

        long elapsed = System.currentTimeMillis() - startTime;
        printResults(totalOrders.get(), totalFills.get(), elapsed, latencySamples);
    }

    private static void seedMarketMakerOrders(MatchingEngine engine) {
        RandomGenerator rng = RandomGenerator.getDefault();
        BigDecimal midPrice = new BigDecimal("150.00");

        System.out.println("Seeding market maker orders...");
        for (int mm = 0; mm < MARKET_MAKER_ACCOUNTS; mm++) {
            String mmId = "mm-" + mm;
            // Quote 5 bid levels and 5 ask levels
            for (int level = 1; level <= 5; level++) {
                BigDecimal bidPrice = midPrice.subtract(
                        new BigDecimal(level * 10).movePointLeft(2));
                BigDecimal askPrice = midPrice.add(
                        new BigDecimal(level * 10).movePointLeft(2));

                // Market maker bids
                engine.submit(Order.create(mmId, SYMBOL, OrderSide.BUY, OrderType.LIMIT,
                        bidPrice, null, 100 + rng.nextInt(400)));
                // Market maker asks
                engine.submit(Order.create(mmId, SYMBOL, OrderSide.SELL, OrderType.LIMIT,
                        askPrice, null, 100 + rng.nextInt(400)));
            }
        }
        System.out.println("Book seeded with " + (MARKET_MAKER_ACCOUNTS * 10) + " resting orders\n");
    }

    private static void runTraderThread(int threadId, MatchingEngine engine, int durationSeconds,
                                        int ordersPerSecond, AtomicLong totalOrders,
                                        AtomicLong totalFills, List<Long> latencySamples) {
        RandomGenerator rng = RandomGenerator.getDefault();
        long endTime = System.currentTimeMillis() + (durationSeconds * 1000L);
        long intervalNs = 1_000_000_000L / ordersPerSecond;

        while (System.currentTimeMillis() < endTime) {
            long loopStart = System.nanoTime();

            Order order = generateRandomOrder(threadId, rng);
            long matchStart = System.nanoTime();
            List<com.titanx.model.Trade> trades = engine.submit(order);
            long latencyMicros = (System.nanoTime() - matchStart) / 1_000;

            totalOrders.incrementAndGet();
            totalFills.addAndGet(trades.size());
            latencySamples.add(latencyMicros);

            // Busy-spin to hit target rate (accurate for benchmarking)
            long elapsed = System.nanoTime() - loopStart;
            if (elapsed < intervalNs) {
                long sleepNs = intervalNs - elapsed;
                try { Thread.sleep(sleepNs / 1_000_000, (int)(sleepNs % 1_000_000)); }
                catch (InterruptedException e) { Thread.currentThread().interrupt(); return; }
            }
        }
    }

    private static Order generateRandomOrder(int threadId, RandomGenerator rng) {
        String accountId = "trader-" + threadId;
        boolean isBuy = rng.nextBoolean();
        OrderSide side = isBuy ? OrderSide.BUY : OrderSide.SELL;

        // Mix of order types: 60% limit, 30% market, 10% IOC
        int typeRoll = rng.nextInt(10);
        OrderType type;
        BigDecimal price;
        if (typeRoll < 6) {
            type = OrderType.LIMIT;
            // Price within 1% of mid
            double deviation = (rng.nextDouble() - 0.5) * 2.0; // -1% to +1%
            price = new BigDecimal(150.0 + deviation)
                    .setScale(2, RoundingMode.HALF_UP);
        } else if (typeRoll < 9) {
            type = OrderType.MARKET;
            price = null;
        } else {
            type = OrderType.IOC;
            price = new BigDecimal("150.00");
        }

        long quantity = 1 + rng.nextLong(99); // 1-100 shares

        return Order.create(accountId, SYMBOL, side, type, price, null, quantity);
    }

    private static void printResults(long totalOrders, long totalFills, long elapsedMs,
                                     List<Long> latencySamples) {
        double throughput = totalOrders * 1000.0 / elapsedMs;
        double fillRate = totalOrders > 0 ? (totalFills * 100.0 / totalOrders) : 0;

        List<Long> sorted = new ArrayList<>(latencySamples);
        Collections.sort(sorted);

        System.out.println("\n=================================================");
        System.out.println("  TitanX Simulation Results");
        System.out.println("=================================================");
        System.out.printf("  Total Orders:    %,d%n", totalOrders);
        System.out.printf("  Total Fills:     %,d%n", totalFills);
        System.out.printf("  Throughput:      %,.1f orders/sec%n", throughput);
        System.out.printf("  Fill Rate:       %.1f%%%n", fillRate);

        if (!sorted.isEmpty()) {
            System.out.println("\n  Matching Latency (µs):");
            System.out.printf("    p50  = %,d µs%n", percentile(sorted, 50));
            System.out.printf("    p95  = %,d µs%n", percentile(sorted, 95));
            System.out.printf("    p99  = %,d µs%n", percentile(sorted, 99));
            System.out.printf("    max  = %,d µs%n", sorted.get(sorted.size() - 1));
        }

        System.out.println("\n=================================================");
        System.out.println("  Interview Talking Points:");
        System.out.println("  - Lock contention: ReentrantReadWriteLock per symbol");
        System.out.println("  - Next step: Disruptor ring buffer for sub-µs latency");
        System.out.println("  - Persistence: async EventStore never blocks hot path");
        System.out.println("=================================================");
    }

    private static long percentile(List<Long> sorted, int pct) {
        int index = (int) Math.ceil(pct / 100.0 * sorted.size()) - 1;
        return sorted.get(Math.max(0, index));
    }
}

File: backend/src/simulation/OrderGenerator.cpp
#include "titanx/engine/MatchingEngine.hpp"
#include "titanx/risk/RiskEngine.hpp"
#include "titanx/risk/CircuitBreaker.hpp"

#include <spdlog/spdlog.h>
#include <thread>
#include <vector>
#include <random>
#include <algorithm>
#include <numeric>
#include <chrono>
#include <atomic>
#include <mutex>
#include <iostream>
#include <iomanip>

/**
 * ============================================================
 * TitanX OrderGenerator — Load Test Simulation (C++17)
 * ============================================================
 *
 * Simulates realistic market activity:
 *   - 10 market makers submitting bid/ask limit orders (60%)
 *   - Retail traders submitting market orders (25%)
 *   - IOC / FOK order mix (15%)
 *
 * Collects per-order latency and reports percentiles.
 */

using namespace titanx;

static constexpr int      NUM_ORDERS       = 100'000;
static constexpr int      NUM_THREADS      = 4;
static constexpr int      ORDERS_PER_THREAD = NUM_ORDERS / NUM_THREADS;
static constexpr int64_t  BASE_PRICE       = price_to_fixed(150.0);   // $150.00
static constexpr int64_t  TICK_SIZE        = price_to_fixed(0.01);    // $0.01

static std::atomic<int> totalTrades{0};
static std::mutex latencyMutex;
static std::vector<long> allLatencies;

void runSimulation(MatchingEngine& engine, int threadId) {
    std::mt19937 rng(42 + threadId);
    std::uniform_real_distribution<> orderTypeDist(0.0, 1.0);
    std::uniform_int_distribution<int64_t> priceDist(-50, 50); // ±50 ticks
    std::uniform_int_distribution<int64_t> qtyDist(1, 500);
    std::uniform_int_distribution<int> sideDist(0, 1);

    std::vector<long> localLatencies;
    localLatencies.reserve(ORDERS_PER_THREAD);

    for (int i = 0; i < ORDERS_PER_THREAD; ++i) {
        double roll = orderTypeDist(rng);
        OrderType type;
        int64_t price = 0;

        if (roll < 0.60) {
            type = OrderType::LIMIT;
            price = BASE_PRICE + priceDist(rng) * TICK_SIZE;
        } else if (roll < 0.85) {
            type = OrderType::MARKET;
        } else if (roll < 0.93) {
            type = OrderType::IOC;
            price = BASE_PRICE + priceDist(rng) * TICK_SIZE;
        } else {
            type = OrderType::FOK;
            price = BASE_PRICE + priceDist(rng) * TICK_SIZE;
        }

        OrderSide side = sideDist(rng) == 0 ? OrderSide::BUY : OrderSide::SELL;
        int64_t qty = qtyDist(rng);
        std::string accountId = "MM-" + std::to_string(threadId);

        Order order = Order::create(accountId, "AAPL", side, type,
                                    price, 0, qty);

        auto start = std::chrono::steady_clock::now();
        auto trades = engine.submit(order);
        auto end = std::chrono::steady_clock::now();

        long latencyMicros = std::chrono::duration_cast<std::chrono::microseconds>(
            end - start).count();
        localLatencies.push_back(latencyMicros);
        totalTrades.fetch_add(static_cast<int>(trades.size()),
                              std::memory_order_relaxed);
    }

    {
        std::lock_guard<std::mutex> lock(latencyMutex);
        allLatencies.insert(allLatencies.end(),
                            localLatencies.begin(), localLatencies.end());
    }
}

int main() {
    spdlog::set_level(spdlog::level::warn); // suppress info during simulation
    spdlog::set_pattern("[%Y-%m-%d %H:%M:%S.%e] [%^%l%$] %v");

    std::cout << "================================================\n"
              << "  TitanX Load Test Simulation (C++)\n"
              << "  " << NUM_ORDERS << " orders, "
              << NUM_THREADS << " threads, AAPL\n"
              << "================================================\n\n";

    // Build engine (no persistence — pure in-memory test)
    CircuitBreaker circuitBreaker;
    RiskEngine riskEngine(circuitBreaker);
    MatchingEngine engine(riskEngine);

    auto startTime = std::chrono::steady_clock::now();

    // Launch threads
    std::vector<std::thread> threads;
    for (int t = 0; t < NUM_THREADS; ++t) {
        threads.emplace_back(runSimulation, std::ref(engine), t);
    }

    for (auto& t : threads) {
        t.join();
    }

    auto endTime = std::chrono::steady_clock::now();
    auto totalMs = std::chrono::duration_cast<std::chrono::milliseconds>(
        endTime - startTime).count();

    // ---- Compute percentiles ----
    std::sort(allLatencies.begin(), allLatencies.end());

    auto percentile = [&](double p) -> long {
        size_t idx = static_cast<size_t>(p * allLatencies.size());
        if (idx >= allLatencies.size()) idx = allLatencies.size() - 1;
        return allLatencies[idx];
    };

    double avgLatency = static_cast<double>(
        std::accumulate(allLatencies.begin(), allLatencies.end(), 0LL))
        / allLatencies.size();

    double throughput = static_cast<double>(NUM_ORDERS) * 1000.0 / totalMs;

    // ---- Report ----
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "Simulation complete!\n\n";
    std::cout << "  Total orders:     " << NUM_ORDERS << "\n";
    std::cout << "  Total trades:     " << totalTrades.load() << "\n";
    std::cout << "  Elapsed time:     " << totalMs << " ms\n";
    std::cout << "  Throughput:       " << throughput << " orders/sec\n\n";

    std::cout << "  Latency Distribution (µs):\n";
    std::cout << "    Min:    " << allLatencies.front() << "\n";
    std::cout << "    P50:    " << percentile(0.50) << "\n";
    std::cout << "    P95:    " << percentile(0.95) << "\n";
    std::cout << "    P99:    " << percentile(0.99) << "\n";
    std::cout << "    P99.9:  " << percentile(0.999) << "\n";
    std::cout << "    Max:    " << allLatencies.back() << "\n";
    std::cout << "    Avg:    " << avgLatency << "\n\n";

    // Book snapshot
    auto snapshot = engine.getSnapshot("AAPL", 5);
    if (snapshot) {
        std::cout << "  Order Book Snapshot (AAPL, top 5 levels):\n";
        std::cout << "    Bids:\n";
        for (const auto& lvl : snapshot->bids) {
            std::cout << "      " << price_to_string(lvl.price)
                      << "  qty=" << lvl.totalQuantity
                      << "  orders=" << lvl.orderCount << "\n";
        }
        std::cout << "    Asks:\n";
        for (const auto& lvl : snapshot->asks) {
            std::cout << "      " << price_to_string(lvl.price)
                      << "  qty=" << lvl.totalQuantity
                      << "  orders=" << lvl.orderCount << "\n";
        }
    }

    std::cout << "\n================================================\n";
    return 0;
}
